tests/unit/crawlers/_http/test_http_crawler.py

```diff
@@ -410,3 +419,62 @@ def mark_request_execution(request: Request) -> Response:  # noqa: ARG001 # Unus
     await crawler.run([Request.from_url(url=test_url)])
 
     assert execution_order == ['pre-navigation-hook 1', 'pre-navigation-hook 2', 'request', 'final handler']
+
+
+@pytest.mark.parametrize(
+    'http_client_class',
+    [
+        pytest.param(CurlImpersonateHttpClient, id='curl'),
+        pytest.param(HttpxHttpClient, id='httpx'),
+    ],
+)
+async def test_isolation_cookies(http_client_class: type[BaseHttpClient], httpbin: URL) -> None:
+    http_client = http_client_class()
+    sessions_ids: list[str] = []
+    sessions_cookies: dict[str, dict[str, str]] = {}
+    response_cookies: dict[str, dict[str, str]] = {}
+
+    crawler = HttpCrawler(
+        session_pool=SessionPool(max_pool_size=1),
+        http_client=http_client,
+        concurrency_settings=ConcurrencySettings(max_concurrency=1),
+    )
+
+    @crawler.router.default_handler
+    async def handler(context: HttpCrawlingContext) -> None:
+        if context.session:
+            sessions_ids.append(context.session.id)
+            if context.request.unique_key in {'1', '2'}:
+                sessions_cookies[context.session.id] = context.session.cookies
+                response_data = json.loads(context.http_response.read())
+                response_cookies[context.session.id] = response_data.get('cookies')
+                if context.request.unique_key == '1':
```

Comment:
Instead of "encoding" this into the unique key, you could use a flag in `user_data`, e.g., `if context.request.user_data.get('retire_session', False)`